File: ruby/bob/bob.rb
class Sentence
  def initialize(words)
    @words = words
  end

  def shouted?
    !lower_case? && upper_case?
  end

  def question?
    words.end_with? '?'
  end

  def silent?
    words.strip.empty?
  end

  private

  attr_reader :words

  def lower_case?
    words == words.downcase
  end

  def upper_case?
    words == words.upcase
  end
end

class Bob
  def hey(words)
    reply_to(Sentence.new(words))
  end

  private

  def reply_to(sentence)
    if sentence.shouted?
      'Woah, chill out!'
    elsif sentence.question?
      'Sure.'
    elsif sentence.silent?
      'Fine. Be that way!'
    else
      'Whatever.'
    end
  end
end

File: elixir/bob/bob.exs
defmodule Teenager do
  def hey(input) do
    cond do
      silence?(input) -> "Fine. Be that way!"
      shout?(input) -> "Woah, chill out!"
      question?(input) -> "Sure."
      true -> "Whatever."
    end
  end
  
  defp silence?(input) do
    String.strip(input) == ""
  end
  
  defp shout?(input) do
    uppercase?(input) and alpha_chars?(input)
  end
  
  defp uppercase?(input) do
    input == String.upcase(input)
  end
  
  defp alpha_chars?(input) do
    String.match? input, ~r/[[:alpha:]]/
  end
  
  defp question?(input) do
    String.ends_with? input, "?"
  end
end
